### PR TITLE
Created a toggle switch to enable and disable groups through the Edit Context

### PR DESCRIPTION
**Pull request by dtamkin1**
Implemented a new ClassBuilder in the EditContext that allows us to pass in a member variable, which is used to create a toggle switch through the PropertyRowWidget and ReflectedPropertyEditor that can enable and disable the group.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.hxx
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.hxx:53]: nit: This is the same as WidgetList, we can just reuse that type here.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp:549]: It would be great if we could slightly expand `Code\Framework\Tests\InstanceDataHierarchy.cpp` to include a test to validate this new "class elements with an associated member" behavior - a test that simply reflects groups with and without an associated member bool variable and validating their layout would be ace!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1141]: nit: arguably easier to read as `m_expanded != checked`

**Review by NicholasVanSickle — COMMENTED** [review]
This is really cool, awesome work! I'd love to move forward with this, but I'd really like to see an InstanceDataHierarchy test case get added before giving this my stamp of approval, as our reflection system is already very complicated & brittle and I'd like to make sure we have a regression test on our backend ready for this change.

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Serialization/EditContext.h:242]: nit: This also needs a closing paren as it seems to be missing here or on the line above

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Serialization/EditContext.h:243]: nit: `to we` -> `so we`, also maybe it should be `serialization` data

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Serialization/EditContext.h:541]: update: This is a little worrying 😅 There should be a portable safe way to do this I'm sure 🙂 Might need to avoid `reinterpret_cast`... at the very least we move this to a small helper function to try and hide this complexity from the main intent of the function? (and maybe try writing a test for it too! 🙂)

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/Serialization/EditContext.h:544]: question: I'm not sure I understand this comment, what is meant by `offset =`?

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp:1115]: nit: Could bump this comment up now the line is quite long (so it's the line above)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1123]: update: We're trying to move away from the old EBus macro calling convention, would you be able to please use the standard template approach instead. I think it should be...

```
PropertyTypeRegistrationMessages::Bus::EventResult(
    m_handler, &PropertyTypeRegistrationMessages::Bus::Events::ResolvePropertyHandler,
    m_handlerName, azrtti_typeid<bool>());
```

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1126]: suggestion: I think this can just be a `static_cast` (or we should use a Qt dynamic_cast to ensure it's the right type)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.hxx
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.hxx:145]: suggestion: We might want a const overload of this too `const QWidget* GetToggle() const { return m_toggleSwitch; }`

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:629]: nit: additional parens are not necessary around the expressions, seems to make the code a little too noisy

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:756]: nit: `(!node->GetElementEditMetadata())` can just be `!node->GetElementEditMetadata()` (additional parens)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:790]: nit: additional parens

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:1388]: suggestion: A guide comment here might be good to help the reader understand why this is needed

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (hultonha) [inline Gems/GradientSignal/Code/Source/GradientSampler.cpp:77]: question: Just checking we definitely want to remove this `ChangeNotify` event?

**Review by hultonha — COMMENTED** [review]
Change looks interesting - I'm not sure I'm a fan of hiding and disabling at the same time (maybe UX can weigh in on this) but definitely seems like a useful feature!

It would be cool to think if there's way to test this (and existing code) to ensure no regressions have occurred (coverage isn't great right now but might be worth giving it some thought about what might be able to be tested in isolation.

Also might be worth doing a run through some Test Rail (manual tests) to check they all still pass (Nick might be able to offer other suggestions related to this).

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (dtamkin1) [reply Code/Framework/AzCore/AzCore/Serialization/EditContext.h:242]: Thanks, will fix!

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (dtamkin1) [reply Code/Framework/AzCore/AzCore/Serialization/EditContext.h:243]: Good catch

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp:1115]: Thanks, will fix.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1123]: Sounds good, I've put in the new EBUS template.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1126]: Sounds good, static_cast works well!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1141]: Thanks, good catch!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:629]: Good catch!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:1388]: Thanks, added it!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.hxx
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.hxx:53]: Sounds good!

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (dtamkin1) [reply Code/Framework/AzCore/AzCore/Serialization/EditContext.h:541]: Thanks, I've moved this code to a helper method that returns DataElement instead

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (dtamkin1) [reply Code/Framework/AzCore/AzCore/Serialization/EditContext.h:544]: See above.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.hxx
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.hxx:145]: Thanks, I've added it.

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (dtamkin1) [reply Gems/GradientSignal/Code/Source/GradientSampler.cpp:77]: Good catch, I've added the ChangeNotify event back.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp
Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/PropertyRowWidget.cpp:1123]: Thanks!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/InstanceDataHierarchy.cpp:549]: I added unit tests for normal groups, toggle groups, and nested groups!

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h
Comment (AMZN-koppersr) [inline Code/Framework/AzCore/AzCore/Serialization/EditContext.h:247]: At face value this seems like something that would be doable through the attributes instead of adding a new element. Are there any restrictions that prevent using attributes to mark groups?

Comment (dtamkin1) [reply Code/Framework/AzCore/AzCore/Serialization/EditContext.h:247]: Well right now there is no current way to have a data element and a group in the same line, and this is the first part of a change that introduces new groupings in the editor: https://jira.agscollab.com/browse/ATOM-14004
Also, the enable/disable logic for the group makes it so that the user doesn't have to do a change notify for each element in the group explicitly.

**Review by NicholasVanSickle — APPROVED** [review]

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (dtamkin1) [reply Gems/GradientSignal/Code/Source/GradientSampler.cpp:77]: The ChangeNotify attribute actually isn't needed for any of the data elements anymore, the property handler is doing the enabling and disabling of each group!

Comment (hultonha) [reply Gems/GradientSignal/Code/Source/GradientSampler.cpp:77]: Okay nice! 🙂

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/Tests/InstanceDataHierarchy.cpp:1450]: update: It's generally preferred to use `EXPECT_...` as opposed to ASSERT macros (so that the rest of the test can run to completion instead of failing there. There's also some great stuff you can use with Matchers to make things more succinct (I'll send you a link to some articles for reference)

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/Tests/InstanceDataHierarchy.cpp:1468]: suggestion: With the naming again, it might be good to say have a test that is called `GroupToggleCanBeExpanded`, and `GroupToggleCanBeCollapsed`, `GroupToggleHidesProperties` etc... When one fails the output is so much more useful. It's hard to do well for sure but can make a massive difference - maybe have a think about how we could break these down a little more

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/Tests/InstanceDataHierarchy.cpp:1503]: suggestion: This code looks almost identical to the code in the other test - it might be worth making a helper function that builds this nodeStack for you (it's fine to have test setup code moved into separate functions so long as it doesn't dramatically impact the readability of the test). Might be worth refactoring this a little bit to remove the duplication

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/Tests/InstanceDataHierarchy.cpp:1517]: suggestion: We're testing quite a lot here in one go, testing closely related functionality is fine, but it might also be worth breaking this up into separate tests (this could be done in a follow-up PR if you'd prefer, but it would make a big difference to the usefulness of the tests (for maintaining and if they fail)

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/Tests/InstanceDataHierarchy.cpp:1567]: update: Same comment as before

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/Tests/InstanceDataHierarchy.cpp:1587]: suggestion: It might be possible to rewrite some of this using matchers too (or even a parameterized test which can remove a lot of boilerplate) - the tests are great but we want to try and treat them the same as we treat production code and make them as clear and compact as we can

**Review by hultonha — COMMENTED** [review]
Massive improvement and excellent to see the tests taking shape! I have a few more comments but this is heading in the right direction and looking really promising 🙂

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:754]: Thanks, I've marked them explicit.

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1415]: Thanks, I changed the namespace!

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1503]: Sounds good, I moved this bit of code to a helper method in the test fixture.

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:754]: Awesome thanks!

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1415]: Amazing thanks! 👍

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:1391]: The `||` is supposed to be here, widgets are never added to both m_widgets and m_specialGroupWidgets, so we want to execute the requestWrite code if we find the appropriate property editor in one of the two maps.

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (dtamkin1) [reply Gems/GradientSignal/Code/Source/GradientSampler.cpp:82]: No problem! In this case, the change notify function was providing a callback, and that callback is being used to refresh everything so that the logic for enable/disable would be called based on the group's boolean field. The cool part about these new toggle groups is that the property handler changes the boolean value and broadcasts the results to each of its data elements that have the ReadOnly attribute.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp
Comment (hultonha) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/PropertyEditor/ReflectedPropertyEditor.cpp:1391]: Okay got you thanks!

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (hultonha) [reply Gems/GradientSignal/Code/Source/GradientSampler.cpp:82]: I see got you, thanks for clarifying! 👍

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1413]: Sounds good, I've changed the test names to make them more descriptive and testing for only one property at a time.

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1587]: I implemented a parameterized test and changed the layout of the other tests to make them more compact and provide greater feedback when they fail :)

Comment (hultonha) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1587]: That's fantastic @dtamkin1, thanks very much for the update! Looking forward to seeing them :+1:

File: Code/Framework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [reply Code/Framework/Tests/InstanceDataHierarchy.cpp:1413]: Fantastic thanks!

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (hultonha) [inline Gems/GradientSignal/Code/Source/GradientSampler.cpp:75]: suggestion: This is super minor and might not be worth doing in this case, but this looks technically like the 'usage' code - sometimes it's nice to split the changes in a PR to the new 'feature/API' that's added and then the code that's actually using it. It is likely overkill here but can be useful to keep in mind for future reference.

File: Code/Framework/AzToolsFramework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [inline Code/Framework/AzToolsFramework/Tests/InstanceDataHierarchy.cpp:1564]: question: I just want to double check here EXPECT_EQ is doing a proper string compare - do you know what this outputs when it fails? (if you update `SubData` to `SubDate` for example just to make sure it fails) - you might be better off using one of the GTest str test functions (though they might only work with std::string types potentially). It's just if you use them you will likely get much better output when the test fails. Worth trying just to be sure but looking much better now

**Review by hultonha — APPROVED** [review]
This looks like a good first pass to me 🙂 I  just have a couple of non-blocking comments but otherwise all good :+1: Thanks for making the updates!

File: Code/Framework/AzToolsFramework/Tests/InstanceDataHierarchy.cpp
Comment (dtamkin1) [reply Code/Framework/AzToolsFramework/Tests/InstanceDataHierarchy.cpp:1564]: Yeah! When the test fails, it prints out the input, the output and the expected value. I found that information very helpful, since if any of these tests were to fail, the developer would have information about exactly which input the test was failing on. This is especially helpful when only some inputs fail. For example, if I change that subChild test to SubDate, then I'll see each of the failing outputs. I'll also see that each of the failing inputs belongs to a sub group, so I'll know to look at the parenting of sub groups.

File: Gems/GradientSignal/Code/Source/GradientSampler.cpp
Comment (dtamkin1) [reply Gems/GradientSignal/Code/Source/GradientSampler.cpp:75]: Okay, thanks I'll keep that in mind for future reference!

**Review by cgalvan — APPROVED** [review]

File: Code/Framework/AzToolsFramework/Tests/InstanceDataHierarchy.cpp
Comment (hultonha) [reply Code/Framework/AzToolsFramework/Tests/InstanceDataHierarchy.cpp:1564]: That's awesome! 😄 Thanks for the info 👍

**Pull request merged**
